tests/csrf_protection_test.py

```diff
@@ -0,0 +1,361 @@
+    form with valid credentials and the retrieved CSRF token
+    to verify successful submission.
+
+    Args:
+        client: The test client for the application.
+        app: The application instance.
+
+    Returns:
+        None
+    """
+    # Get the login page to retrieve the CSRF token
+    response = client.get("/login")
+    csrf_token = get_csrf_token_from_response(response.data.decode())
+
+    # Submit the form with CSRF token
+    response = client.post(
+        "/login",
+        data={
+            "email": "user@example.com",
+            "password": "Str0ngP@$$word123!C0ntre$namUyfue&t3",
+            "csrf_token": csrf_token,
+        },
+        follow_redirects=True,
+    )
+    assert (
+        response.status_code == 200
+    ), "Form should accept submission with valid CSRF token"
+
+
+def test_login_post_invalid_csrf(client):
+    """Attempts to submit the login form with an invalid CSRF
+    token to verify that the form correctly rejects the submission.
+
+    Args:
+        client: The test client for the application.
+
+    Returns:
+        None
+    """
+    response = client.post(
+        "/login",
+        data={
+            "email": "test@example.com",
+            "password": "TestP@ssword!4234m!@3",
+            "csrf_token": "invalid_csrf_token",
+        },
+        follow_redirects=True,
+    )
+    assert (
+        "Invalid CSRF token. Please try again." in response.data.decode()
+        or response.status_code == 400
+    ), "Form should reject submission with invalid CSRF token"
+
+
+def test_login_post_valid_csrf(client, test_user):
+    """Submits the login form with a valid CSRF token
+    to verify that the form accepts the submission.
+
+    Args:
+        client: The test client for the application.
+        test_user: The test user instance.
+
+    Returns:
+        None
+    """
+    response = client.get("/login")
+    csrf_token = get_csrf_token_from_response(response.data.decode())
+
+    response = client.post(
+        "/login",
+        data={
+            "email": "test@example.com",
+            "password": "TestP@ss3fsadf3!@!@#",
+            "csrf_token": csrf_token,
+        },
+        follow_redirects=True,
+    )
+    assert (
+        response.status_code == 200
+    ), "Form should accept submission with valid CSRF token"
+
+
+def test_signup_form_rejection_without_csrf(client):
+    """Verifies that submitting the signup form
+    without a CSRF token results in a rejection.
+
+    Args:
+        client: The test client for the application.
+
+    Returns:
+        None
+    """
+    response = client.post(
+        "/signup",
+        data={
+            "email": "test@example.com",
+            "password": "TestP@ss35!@!@#daf3",
+            "password_confirm": "TestP@ss35!@!@#daf3",
+        },
+        follow_redirects=True,
+    )
+    assert (
+        "The CSRF token is missing." in response.data.decode()
+        or response.status_code == 400
```

Comment:
Check this link: https://docs.python.org/3/library/http.html 

we can use:

`response.status_code == HTTPStatus.BAD_REQUEST`